Implement a subset of the Common Workflow Language

nsoranzo opened the pull request:
Adding support for a large subset of CWL [v1.0.2](https://www.commonwl.org/v1.0/), [v1.1](https://www.commonwl.org/v1.1/) and [v1.2](https://www.commonwl.org/v1.2/) . This is a group work, led by @jmchilton with contributions from @mr-c , @mvdbeek , @hmenager and myself.
This effort wouldn't have been possible without the support of [ELIXIR](https://elixir-europe.org/) which sponsored this project at [BioHackathon Europe](https://biohackathon-europe.org/) in 2018, 2019, 2020 and 2021.

CWL Support (Tools):
--------------------

- Implemented integer, long, float, double, boolean, string, File, Directory, "null", Any, as well as records and arrays thereof. There are two approaches to handling more complex parameters discussed here (https://github.com/common-workflow-lab/galaxy/issues/59).
- ``secondaryFiles`` that are actual Files are implemented, secondaryFiles containing directories are not yet implemented.
- ``InlineJavascriptRequirement`` are support to define output files (see ``test_cat3`` test case).
- ``EnvVarRequirement``s are supported (see the ``test_env_tool1`` and ``test_env_tool2`` test cases).
- Expression tools are supported (see ``parseInt-tool`` test case).
- Shell tools are also support (see record output test case).
- Default File values are very un-Galaxy and have been hacked into work with Tools - they still don't work with workflows.
- Partial Docker support - this supports the most simple and common pullFrom semantics but not additional ways to fetch containers or additional options such as output directory configuration (https://github.com/common-workflow-lab/galaxy/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20Docker). Additionally, Galaxy mounts the inputs and outputs where it wants instead of CWL required mount points - this needs to be fixed for the conformance tests but may not matter much in practice (I'm not sure).

CWL Support (Workflows):
------------------------

- Simple connections and tool execution.
- Overriding tool input defaults via literal values and simple expressions.
- MultipleInputFeatureRequirements to glue together multiple file inputs into a File[] or multiple File[] into a single flat File[]. (nested merge is still a TODO).
- Simple scatter semantics for Files and non-Files (e.g. count-lines3).
- Simple subworkflows (e.g. count-lines10).
- Simple valueFrom expressions (e.g. ``step-valueFrom`` and ``step-valueFrom2``). This work doesn't yet model non-tool parameters to steps - for complex ``valueFrom`` expressions like in ``step-valueFrom3`` do not work yet.

Remaining Work
---------------------------------

The work remaining is vast and will be tracked at https://github.com/common-workflow-lab/galaxy/issues for the time being.

Implementation Notes:
----------------------

Tools:

 - Non-File CWL outputs are represented as ``expression.json`` files. Traditionally Galaxy hasn't supported non-File outputs from tools but CWL Galaxy has work in progress on bringing native Galaxy support for such outputs #27.
 - CWL secondary files are just normal datasets with extra files stored in ``__secondary_files__`` directory in the dataset's extra_files_path directory and indexed in a file called __secondary_files_index.json in extra_files_path. The upload tools has been augmented to allow attaching arbitrary extra files as a tar file to support getting data into this format initially. CWL requires staging files to include their parent File's ``basename`` - but tools describe inputs as just the extension. I'm not sure which way Galaxy should store __secondary_files__ in its objectstore - just with the extension or with the basename and extension - both options are implemented and can be swapped by setting the boolean STORE_SECONDARY_FILES_WITH_BASENAME in galaxy.tools.cwl.util.
 - CWL Directory types are datasets of a new type "directory" implemented earlier in this branch.
 - The tool execution API has been extended to add a ``inputs_representation`` parameter that can be set to "cwl" now. The ``cwl`` representation for running tools corresonding to the CWL job json format with {class: "File: path: "/path/to/file"} inputs replaced with {"src": "hda", "id": "<dataset_id>"}. Code for building these requests for CWL job json is available in the test class.
 - Since the CWL <-> Galaxy parameter translation may change over time, for instance if Galaxy develops or refines parameter classes - CWL state and CWL state version is tracked in the database and hopefully for reruns, etc... we could update the Galaxy state from an older version to a new one.
 - CWL allows output parameters to be either ``File`` or non-``File`` and determined at runtime, so ``galaxy.json`` is used to dynamically adjust output extension as needed for non-``File`` parameters.

Workflows:

- This work serializes embedded and referenced tools into the database - this will allow reuse and tracing without require the path to exist forever on the filesystem - this will have problems with default file references in workflows.
- Implements re-mapping CWL workflow connections to Galaxy input connections.
- Fix tool serialization for jobs for path-less tools (such as embedded tools).
- Hack tool state during workflow import for CWL.
- The sort of dynamic shaping of inputs CWL allows has required enhancing Galaxy's map/reduce stuff to allow mapping over dynamic collections that don't yet exist at the time of tool execution and need to be created on the fly. This commit creates them as HDCAs - but likely they should be something else that doesn't appear in the history panel.
- Multi-input scattering but only scatterMethod == "dotproduct" is currently support. Other scatter methods (nested_crossproduct and flatcross_product) are not used by workflows in GA4GH challenge.

Implementation Description:
-----------------------------

The [reference implementation Python library](https://github.com/common-workflow-language/cwltool) (mainly developed by Peter Amstutz) is used to load tool files ending with ``.json`` or ``.cwl`` and proxy objects are created to adapt these tools to Galaxy representations. In particular input and output descriptions are loaded from the tool.

When the tool is submitted, a special specialized tool class is used to build a cwltool compatible job description from the supplied Galaxy inputs and the CWL reference implementation is used to generate a CWL reference implementation Job object. A command-line is generated from this Job object.

As a result of this - Galaxy largely does not need to worry about the details of command-line adapters, expressions, etc....

Galaxy writes a description of the CWL job that it can reload to the job working directory. After the process is complete (on the Galaxy compute server, but outside the Docker container) this representation is reloaded and the dynamic outputs are discovered and moved to fixed locations as expected by Galaxy. CWL allows for much more expressive output locations than Galaxy, for better or worse, and this step uses cwltool to adapt CWL to Galaxy outputs.

Currently all ``File`` outputs are sniffed to determined a Galaxy datatype, CWL allows refinement on this and this remains work to be done.

  1) CWL should support EDAM declaration of types and Galaxy should provide a mapping to core datasets to skip sniffing is types are found.
  2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)

Implementation Links:
----------------------

Hundreds of commits have been rebased into this one and so the details of individual parts of the implementation and how they built on each other are not entirely clear. To see the original ideas behind individual features - here are some relevant links:

- Implement merge_nested link semantics for workflow steps (a903abd).
- Implement subworkflows in CWL (9933c3c)
- MultipleInputFeatureRequirements:
  - Second attempt: ed8307f
  - First attempt: ae11f56
- Basic, implicit dotproduct scattering of workflows - d1ad64e.
- Simple input StepInputExpressionRequirements - 819a27b
- StepInputExpressionRequirements for multiple inputs - 5e7f622
- Record Types in CWL - e6be28a
- Rework original approach at mapping CWL state to tool state - 669ea55
- Rework approach at mapping CWL state to tool state again to use "FieldTypeToolParameter"s - implements default values, optional parameters, and union types for workflow inputs. d1ca22f
- Initial tracking of "cwl_filename" for CWL jobs (67ffc55).
- Reworked secondary file staging, implement testing and indexing of secondary files - 03d1636.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

jmchilton committed:
WIP: Work toward Galaxy-flavored CWL tools.

gx:Interface.

Co-authored-by: John Chilton <jmchilton@gmail.com>
Co-authored-by: jra001

jmchilton committed:
CWL Testing and Runner Improvements.

jmchilton committed:
[WIP] Implement client UI for field parameter type for CWL.

[29 Oct 23] Had to remove because we're not using backbone anymore

jmchilton committed:
Deal with workflow definitions without position fields.

jmchilton committed:
Implement subset of the Common Workflow Language tool and workflow formats.

This should support a subset of [draft-3](http://www.commonwl.org/draft-3/) and [v1.0](http://www.commonwl.org/v1.0/) tools.

CWL Support (Tools):
--------------------

- Implemented integer, long, float, double, boolean, string, File, Directory, "null", Any, as well as records and arrays thereof. There are two approaches to handling more complex parameters discussed here (https://github.com/common-workflow-language/galaxy/issues/59).
- ``secondaryFiles`` that are actual Files are implemented, secondaryFiles containing directories are not yet implemented.
- ``InlineJavascriptRequirement`` are support to define output files (see ``test_cat3`` test case).
- ``EnvVarRequirement``s are supported (see the ``test_env_tool1`` and ``test_env_tool2`` test cases).
- Expression tools are supported (see ``parseInt-tool`` test case).
- Shell tools are also support (see record output test case).
- Default File values are very un-Galaxy and have been hacked into work with Tools - they still don't work with workflows.
- Partial Docker support - this supports the most simple and common pullFrom semantics but not additional ways to fetch containers or additional options such as output directory configuration (https://github.com/common-workflow-language/galaxy/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20Docker). Additionally, Galaxy mounts the inputs and outputs where it wants instead of CWL required mount points - this needs to be fixed for the conformance tests but may not matter much in practice (I'm not sure).

CWL Support (Workflows):
------------------------

- Simple connections and tool execution.
- Overriding tool input defaults via literal values and simple expressions.
- MultipleInputFeatureRequirements to glue together multiple file inputs into a File[] or multiple File[] into a single flat File[]. (nested merge is still a TODO).
- Simple scatter semantics for Files and non-Files (e.g. count-lines3).
- Simple subworkflows (e.g. count-lines10).
- Simple valueFrom expressions (e.g. ``step-valueFrom`` and ``step-valueFrom2``). This work doesn't yet model non-tool parameters to steps - for complex ``valueFrom`` expressions like in ``step-valueFrom3`` do not work yet.

Remaining Work
---------------------------------

The work remaining is vast and will be tracked at https://github.com/common-workflow-language/galaxy/issues for the time being.

Implementation Notes:
----------------------

Tools:

 - Non-File CWL outputs are represented as ``expression.json`` files. Traditionally Galaxy hasn't supported non-File outputs from tools but CWL Galaxy has work in progress on bringing native Galaxy support for such outputs https://github.com/common-workflow-language/galaxy/issues/27.
 - CWL secondary files are just normal datasets with extra files stored in ``__secondary_files__`` directory in the dataset's extra_files_path directory and indexed in a file called __secondary_files_index.json in extra_files_path. The upload tools has been augmented to allow attaching arbitrary extra files as a tar file to support getting data into this format initially. CWL requires staging files to include their parent File's ``basename`` - but tools describe inputs as just the extension. I'm not sure which way Galaxy should store __secondary_files__ in its objectstore - just with the extension or with the basename and extension - both options are implemented and can be swapped by setting the boolean STORE_SECONDARY_FILES_WITH_BASENAME in galaxy.tools.cwl.util.
 - CWL Directory types are datasets of a new type "directory" implemented earlier in this branch.
 - The tool execution API has been extended to add a ``inputs_representation`` parameter that can be set to "cwl" now. The ``cwl`` representation for running tools corresonding to the CWL job json format with {class: "File: path: "/path/to/file"} inputs replaced with {"src": "hda", "id": "<dataset_id>"}. Code for building these requests for CWL job json is available in the test class.
 - Since the CWL <-> Galaxy parameter translation may change over time, for instance if Galaxy develops or refines parameter classes - CWL state and CWL state version is tracked in the database and hopefully for reruns, etc... we could update the Galaxy state from an older version to a new one.
 - CWL allows output parameters to be either ``File`` or non-``File`` and determined at runtime, so ``galaxy.json`` is used to dynamically adjust output extension as needed for non-``File`` parameters.

Workflows:

- This work serializes embedded and referenced tools into the database - this will allow reuse and tracing without require the path to exist forever on the filesystem - this will have problems with default file references in workflows.
- Implements re-mapping CWL workflow connections to Galaxy input connections.
- Fix tool serialization for jobs for path-less tools (such as embedded tools).
- Hack tool state during workflow import for CWL.
- The sort of dynamic shaping of inputs CWL allows has required enhancing Galaxy's map/reduce stuff to allow mapping over dynamic collections that don't yet exist at the time of tool execution and need to be created on the fly. This commit creates them as HDCAs - but likely they should be something else that doesn't appear in the history panel.
- Multi-input scattering but only scatterMethod == "dotproduct" is currently support. Other scatter methods (nested_crossproduct and flatcross_product) are not used by workflows in GA4GH challenge.

Implementation Description:
-----------------------------

The reference implementation Python library (mainly developed by Peter Amstutz - https://github.com/common-workflow-language/common-workflow-language/tree/master/reference) is used to load tool files ending with ``.json`` or ``.cwl`` and proxy objects are created to adapt these tools to Galaxy representations. In particular input and output descriptions are loaded from the tool.

When the tool is submitted, a special specialized tool class is used to build a cwltool compatible job description from the supplied Galaxy inputs and the CWL reference implementation is used to generate a CWL reference implementation Job object. A command-line is generated from this Job object.

As a result of this - Galaxy largely does not need to worry about the details of command-line adapters, expressions, etc....

Galaxy writes a description of the CWL job that it can reload to the job working directory. After the process is complete (on the Galaxy compute server, but outside the Docker container) this representation is reloaded and the dynamic outputs are discovered and moved to fixed locations as expected by Galaxy. CWL allows for much more expressive output locations than Galaxy, for better or worse, and this step uses cwltool to adapt CWL to Galaxy outputs.

Currently all ``File`` outputs are sniffed to determined a Galaxy datatype, CWL allows refinement on this and this remains work to be done.

  1) CWL should support EDAM declaration of types and Galaxy should provide a mapping to core datasets to skip sniffing is types are found.
  2) For finer grain control within Galaxy, extensions to CWL should allow setting actual Galaxy output types on outputs. (Distinction between fastq and fastqsanger in Galaxy is very important for instance.)

Implementation Links:
----------------------

Hundreds of commits have been rebased into this one and so the details of individual parts of the implementation and how they built on each other are not enitrely clear. To see the original ideas behind individual features - here are some relevant links:

- Implement merge_nested link semantics for workflow steps (https://github.com/common-workflow-language/galaxy/commit/a903abdb65e54d11d1f25d732f9e04da15cc1ff6).
- Implement subworkflows in CWL (https://github.com/common-workflow-language/galaxy/commit/9933c3c211b93b328cf63afba66f144842ff9e10)
- MultipleInputFeatureRequirements:
  - Second attempt: https://github.com/common-workflow-language/galaxy/commit/ed8307f5c3a5e8ea4318807c893543c45fa72cf7
  - First attempt: https://github.com/common-workflow-language/galaxy/commit/ae11f5662cd5254bde1c1cb345d780be8317f59b
- Basic, implicit dotproduct scattering of workflows - https://github.com/common-workflow-language/galaxy/commit/d1ad64ed6f2e000ae1d261c7c22f54ab3b9dcb4d.
- Simple input StepInputExpressionRequirements - https://github.com/common-workflow-language/galaxy/commit/819a27b816ecc71c2190efb33b9127658667e784
- StepInputExpressionRequirements for multiple inputs - https://github.com/common-workflow-language/galaxy/commit/5e7f622b4ff072eafa844af525f4367de5e7e644
- Record Types in CWL - https://github.com/common-workflow-language/galaxy/commit/e6be28a2d4983184757aafd1fafdb1b56b692438
- Rework original approach at mapping CWL state to tool state - https://github.com/common-workflow-language/galaxy/commit/669ea55a7ab55e8741a82d0dcd5a15f5075358dd
- Rework approach at mapping CWL state to tool state again to use "FieldTypeToolParameter"s - implements default values, optional parameters, and union types for workflow inputs. https://github.com/common-workflow-language/galaxy/commit/d1ca22fab5ca8adb8691b2f76846c41aaa798e49
- Initial tracking of "cwl_filename" for CWL jobs (https://github.com/common-workflow-language/galaxy/commit/67ffc5573f45d8293f9a4df502c56247964e11ab).
- Reworked secondary file staging, implement testing and indexing of secondary files - https://github.com/common-workflow-language/galaxy/commit/03d1636fc2eb7ba4e5b42526ea69cce447252c46.

Testing:
---------------------

    % git clone https://github.com/common-workflow-language/galaxy.git
    % cd galaxy
    % git checkout cwl-1.0

Start Galaxy.

    % GALAXY_RUN_WITH_TEST_TOOLS=1 sh run.sh

Open http://localhost:8080/ and see CWL test tools (along with all Galaxy test tools) in left hand tool panel.

To go a step further and actually run CWL jobs within their designated Docker containers, copy the following minimal Galaxy job configuration file to ``config/job_conf.xml``. (Adjust the ``docker_sudo`` parameter based on how you execute Docker).

https://gist.github.com/jmchilton/3997fa471d1b4c556966

Run API tests demonstrating the various CWL demo tools with the following command.

```
./run_tests.sh -api test/api/test_tools_cwl.py
./run_tests.sh -api test/api/test_workflows_cwl.py
./run_tests.sh -api test/api/test_cwl_conformance_v1_0.py
```

An individual conformance test can be ran using this pattern:

```
./run_tests.sh -api test/api/test_cwl_conformance_v1_0.py:CwlConformanceTestCase.test_conformance_v1_0_6
```

The first two execute various tool and workflow test cases manually crafted during implementation of this work. The third is an auto-generate test case class that contains Python tests for every CWL conformance test found with the reference specification.

Issues and Contact
---------------------------------

Report issues at https://github.com/common-workflow-language/galaxy/issues and feel free ping jmchilton on the CWL [Gitter channel](https://gitter.im/common-workflow-language/common-workflow-language).

Co-authored-by: Hervé MENAGER <herve.menager@pasteur.fr>
Co-authored-by: John Chilton <jmchilton@gmail.com>
Co-authored-by: Michael R. Crusoe <michael.crusoe@gmail.com>
Co-authored-by: jra001k
Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>
Co-authored-by: Nicola Soranzo <nicola.soranzo@earlham.ac.uk>

jmchilton committed:
WORKAROUND TO GET TAR TO DIRECTORY WORKING AGAIN.

mvdbeek committed:
Fix non_data_connection workflows

mvdbeek committed:
Fix handling of uploaded_file_name

mvdbeek committed:
Fix directory location tests

mr-c committed:
start documenting state of CWL support

mvdbeek committed:
Add sentinel value workaround for GALAXY_SLOTS hack

davelopez commented on lib/galaxy/schema/schema.py:
```diff
@@ -999,6 +999,11 @@ class CreateNewCollectionPayload(Model):
+        description="List of fields to create for this collection. Set to 'auto' to guess fields from identifiers.",
+        alias="fields_",
+    )
 
```
I think what you need here is the opposite, so there is no conflict with the internal `fields` property of `BaseModel`.
```suggestion
    fields_: Optional[Union[str, List[Dict['str', 'str']]]] = Field(
        default=[],
        description="List of fields to create for this collection. Set to 'auto' to guess fields from identifiers.",
        alias="fields",
    )
```

mvdbeek committed:
Assert length of input connections, instead of inputs when disconnecting steps

Inputs become runtime inputs due to:
```
        if "inputs" in step_dict:
            for input_dict in step_dict["inputs"]:
                step_input = model.WorkflowStepInput(step)
                ...
```
in galaxy.managers.workflows.WorkflowContentsManager.

I guess it is debatable whether we want to adjust the test
or WorkflowContentsManager for disconnected steps.
When exported to .ga these are exactly the same (minus uuid
differences).

mr-c committed:
Fix type hints

jmchilton commented on lib/galaxy/job_execution/output_collect.py:
```diff
@@ -638,15 +638,16 @@ def default_exit_code_file(files_dir, id_tag):
 def collect_extra_files(object_store, dataset, job_working_directory):
     file_name = dataset.dataset.extra_files_path_name_from(object_store)
-    temp_file_path = os.path.join(job_working_directory, "working", file_name)
+    output_dir = "working" if os.path.exists(os.path.join(job_working_directory, "working", file_name)) else "outputs"
```
I'd feel more comfortable about this if we were choosing between the two based on the job type in some way.

mr-c commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -249,9 +249,8 @@ def wrap_input(input_values, input):
-                if value is None:
-                    field_wrapper = None
-                else:
+                field_wrapper: Optional[Union[InputValueWrapper, DatasetFilenameWrapper]] = None
```
I would remove the `field_wrapper` definition from here and move it to the first usage after ` if src == "json"` but without the `Optional`

``` python
if src == "json":
                        field_wrapper: Union[InputValueWrapper, DatasetFilenameWrapper] = InputValueWrapper(input, value, param_dict)
```

as the exception afterwards guarantees that `field_wrapper` will never be `None`

nsoranzo replied:
I don't think we can remove the definition because `field_wrapper = None` is still used on line 266 (`input_values[input.name] = field_wrapper`) if `value` is `None`.

nsoranzo commented on lib/galaxy/app_unittest_utils/tools_support.py:
```diff
@@ -83,14 +83,18 @@ def _init_tool(
     ):
+        if tool_path is None:
+            self.tool_file = os.path.join(self.test_directory, filename)
+            contents_template = string.Template(tool_contents or SIMPLE_TOOL_CONTENTS)
```
```suggestion
            contents_template = string.Template(tool_contents)
```

nsoranzo commented on lib/galaxy_test/base/populators.py:
```diff
@@ -221,6 +235,297 @@ def _raise_skip_if(check, *args):
+                tool_ids = [itemgetter("id")(_) for _ in tools]
+                if raw_tool_id in tool_ids:
+                    galaxy_tool_id = raw_tool_id
+                    tool_uuid = None
```
```suggestion
```

nsoranzo commented on lib/galaxy_test/base/populators.py:
```diff
@@ -488,6 +793,10 @@ def run_tool_payload(self, tool_id: str, inputs: dict, history_id: str, **kwds)
                 kwds["__files"][key] = value
                 del inputs[key]
 
+        ir = kwds.get("inputs_representation", None)
```
```suggestion
        ir = kwds.get("inputs_representation")
```

nsoranzo commented on run_tests.sh:
```diff
@@ -404,6 +405,7 @@ do
           test_script="pytest"
           report_file="run_framework_tests.html"
           framework_test=1;
+          generate_cwl_conformance_tests=1
```
Is it actually needed for framework tests?

mvdbeek replied:
Since we don't add these to the samples_tool_conf.xml file I don't think they are

nsoranzo replied:
We do add some of them, the ones used in API tests: https://github.com/galaxyproject/galaxy/pull/12909/files#diff-a4c6a629466aed28ef86982600acabb5642b9c7247318530cea6fd2cc2baabb5

mvdbeek replied:
~~Ah, ok, then we need at least those tools.~~
But I guess the framework tests won't load CWL tools by default ... I guess it'd be a little noisy on startup, but no further harm if they are not available ?

nsoranzo commented on lib/galaxy/tool_util/parser/cwl.py:
```diff
@@ -158,7 +253,23 @@ def parse_requirements_and_containers(self):
+            if cores_min:
+                cores_min = int(cores_min)
+                if cores_min < 1:
+                    raise ValueError("cores_min must be a positive integer")
```
This can be dynamic in CWL, e.g. in `dynresreq.cwl` :

```yaml
requirements:
  ResourceRequirement:
      ramMin: 8
      coresMin: $(inputs.special_file.size)
      coresMax: $(inputs.special_file.size)

inputs:
  special_file: File
```

mvdbeek committed:
Disable cheetah in configfiles, env vars for CWL tools

mvdbeek committed:
Drop test_deserialize_cwl_tool, already testing that more accurately in cwl_test.py

mvdbeek committed:
Fix wrong resolution of Any type when re-using CWL tools

Calling job() modifies the internal state, so running
`test_conformance_v1_1_clt_any_input_with_file_provided` followed by
`test_conformance_v1_1_clt_any_input_with_integer_provided` fails
because Galaxy infers that we're dealing with a dataset that needs to be
wrapped, instead of the integer we're really dealing with.

Can be recitulated with `./run_tests.sh -cwl 'lib/galaxy_test/api/cwl/test_cwl_conformance_v1_2.py' -- -m green -k test_conformance_v1_1_clt_any_input`

mvdbeek committed:
Coerce discovered optional files to data

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -76,6 +77,23 @@ def create_tool(self, trans, tool_payload, allow_load=True):
+            uuid = tool_payload.get("uuid") or representation.get('uuid')
+            if uuid is None:
+                uuid = str(uuid4())
+            tool_id = representation.get("id", None)
```
This gets overwritten in the following if/else

```suggestion
```

mvdbeek committed:
Fix complex types via record collection type

mvdbeek committed:
Fix handle_known_output for nested output records

mvdbeek committed:
Skip staging inputs for outputs

mvdbeek committed:
Fix packed document support if main/#main is tool instead of workflow

nsoranzo commented on lib/galaxy/managers/executables.py:
```diff
@@ -8,6 +8,8 @@
     object_id = as_dict.get("object_id", None)
     if as_dict.get("src", None) == "from_path":
+        # FIXME: assert trans is not None, but probably breaks conformance tests
+        # Don't merge before we haven't addressed this
```
This is actually already in since 20.01, not sure how much it's used.

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -76,6 +78,25 @@ def create_tool(self, trans, tool_payload, allow_load=True):
+                # hash.
+                proxy = tool_proxy(tool_object=representation['raw_process_reference'], tool_directory=tool_directory, uuid=uuid)
+            tool_id = proxy.galaxy_id()
+            value = representation
```
```suggestion
```

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -95,7 +117,7 @@ def create_tool(self, trans, tool_payload, allow_load=True):
                 uuid=uuid,
                 value=value,
```
```suggestion
                value=representation,
```

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -76,6 +78,25 @@ def create_tool(self, trans, tool_payload, allow_load=True):
 
             tool_version = representation.get("version", None)
             value = representation
```
```suggestion
```

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -76,6 +78,25 @@ def create_tool(self, trans, tool_payload, allow_load=True):
+            uuid = tool_payload.get("uuid") or representation.get('uuid')
+            if uuid is None:
+                uuid = str(uuid4())
+            tool_version = representation.get("version", None)
```
```suggestion
```

nsoranzo commented on lib/galaxy/managers/tools.py:
```diff
@@ -84,6 +105,7 @@ def create_tool(self, trans, tool_payload, allow_load=True):
+            assert existing_tool.uuid == uuid
             dynamic_tool = existing_tool
         else:
             dynamic_tool = self.create(
```
```suggestion
            tool_version = representation.get("version")
            dynamic_tool = self.create(
```

mvdbeek committed:
Implement rough mapping between EDAM formats and datatypes

This adds the cwl_format parameter to the upload tool and exposes the
cwl_formats field in the details view of the HDA API response, which
contains a list of URIs for the format. This also provides the
edam_format from the datatype annotation as the format field of the cwl
job input description and uses the format field within the cwl job
outputs to write out a gaalxy.json file that contains the cwl-format,
which we then use to pick a datatype (if we find one).

Not sure this is a good idea, in practice this will fall apart for
compressed outputs at least, for which there is no distinction in edam.
If that's the only problem we could probably work around the issue, but
there could be more problems. It is kind of cool and maybe necessary for
interoperability between cwl tool outputs and galaxy tool inputs
(assuming users won't want to change dataypes), but maybe we should just always
stick the edam format into the hda metadata or into a separate hda
column and keep these two concepts separate.

We should maybe extend the edam format use a bit more. This could be
a cool way for tools to use user-defined datatypes that Galaxy doesn't
(yet) know about.

mvdbeek committed:
Support uploading directory literals

mvdbeek committed:
Keep directory parameters in job parameters

and set LoadListingRequirement as supported requirement, since cwltool
handles this for us.

nsoranzo commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -1592,8 +1596,9 @@ def _finish_dataset(self, output_name, dataset, job, context, final_job_state, r
-            if dataset.ext == 'auto':
-                dataset.extension = context.get('ext', 'data')
+            context_ext = context.get('ext', 'data')
+            if dataset.ext == 'auto' or dataset.ext == 'data' and context_ext != 'data':
```
Parentheses would make this more readable.

nsoranzo commented on lib/galaxy/jobs/runners/util/job_script/DEFAULT_JOB_FILE_TEMPLATE.sh:
```diff
@@ -35,6 +35,7 @@ GALAXY_LIB="$galaxy_lib"
 _galaxy_setup_environment "$PRESERVE_GALAXY_ENVIRONMENT"
 export _GALAXY_JOB_HOME_DIR
 export _GALAXY_JOB_TMP_DIR
+_GALAXY_PATH="$PATH"
```
I think this can be dropped.

nsoranzo commented on lib/galaxy/tool_util/client/staging.py:
```diff
@@ -142,6 +142,7 @@ def _attach_file(upload_payload, uri, index=0):
                 )
                 tags = upload_target.properties.get("tags")
                 fetch_payload["targets"][0]["elements"][0]["tags"] = tags
+            print("\n\n\n\n\n\nFETCH_PAYLOAD IS %s\n\n\n\n\n" % fetch_payload)
```
```suggestion
```

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -3000,6 +3046,170 @@ def job_failed(self, job_wrapper, message, exception=False):
+        raise NotImplementedError()
+
+
+class GalacticCwlTool(CwlCommandBindingTool):
+    """A CWL tool with a gx:Interface defined so Galaxy tool state can be used."""
+
+    tool_type = "galactic_cwl"
+
+    def param_dict_to_cwl_inputs(self, param_dict, local_working_directory):
+        from galaxy.tool_util.cwl.representation import galactic_flavored_to_cwl_job
+
+        input_json = galactic_flavored_to_cwl_job(self, param_dict, local_working_directory)
+        return input_json
+
+
```
Do we still want to add `GalacticCwlTool` or should we drop this?

mvdbeek replied:
Do we have to make a decision here ? I think it's a good idea, but if we remove this we should put it into a separate commit (if it isn't already)

nsoranzo replied:
Do you mean remove it in a separate commit?

mvdbeek replied:
I mean to refactor the parts needed into a single commit that we can have somewhere to apply later.

mr-c commented on doc/source/dev/cwl.md:
```diff
@@ -0,0 +1,8 @@
+CWL import in Galaxy
+
```
```suggestion
1. List paths to CWL tools in `tool_conf.xml`.
2. Set the following in  `galaxy.yml`: 
enable_beta_tool_formats: true
enable_beta_workflow_modules: true
enable_beta_workflow_format: true
check_upload_content: false
strict_cwl_validation: false
```

nsoranzo replied:
@mvdbeek Is `enable_beta_workflow_format` actually needed? Can't find any related code.

mvdbeek replied:
It might not be, I just copied this from the test script.

nsoranzo commented on scripts/end_cwl_conformance.bash:
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
```
Can we drop this script?

nsoranzo commented on scripts/run_cwl.py:
```diff
@@ -0,0 +1,92 @@
+    args = arg_parser.parse_args(argv)
+
+    gi = galaxy.GalaxyInstance(args.host, args.api_key)
+    i = GiPostGetMixin()
```
```suggestion
    i = GiHttpMixin()
```

nsoranzo commented on scripts/run_cwl.py:
```diff
@@ -0,0 +1,92 @@
+from galaxy_test.base.populators import (  # noqa: I100,I202
+    CwlPopulator,
+    GiDatasetPopulator,
+    GiPostGetMixin,
```
```suggestion
    GiHttpMixin,
```

nsoranzo commented on scripts/run_cwl_conformance.py:
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python
```
Can we drop this script?

nsoranzo commented on scripts/run_for_cwl_conformance.bash:
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Can we drop this script?

nsoranzo commented on scripts/run_for_cwl_conformance.bash:
```diff
@@ -0,0 +1,41 @@
+export GALAXY_CONFIG_OVERRIDE_ADMIN_USERS="cwluser@example.com"
+
+export GALAXY_CONFIG_OVERRIDE_ENABLE_BETA_WORKFLOW_MODULES="true"
+export GALAXY_CONFIG_OVERRIDE_ENABLE_BETA_WORKFLOW_FORMAT="true"
```
```suggestion
```

mvdbeek committed:
Merge subworkflow input logic?

mvdbeek committed:
Drop divergent to_cwl/from_cwl, factor out extra_step_state building

mvdbeek committed:
TreeDict fix

mvdbeek committed:
Use regular staging for CWL tests instead of allow_path_paste, which is broken

mvdbeek committed:
Fix directory uploads

mvdbeek committed:
Record unnamed_outputs as job outputs

mvdbeek committed:
Download complex outputs

mvdbeek committed:
Download secondary files as well

mr-c commented on lib/galaxy/tool_util/cwl/util.py:
```diff
@@ -589,6 +625,7 @@ def dir_listing(dir_path):
                     "basename": basename,
                     "listing": listing,
                 }
+                _handle_pseudo_location(properties, pseudo_location=pseudo_location)
```
> `TypeError: _handle_pseudo_location() missing 1 required positional argument: 'download_url'`

nsoranzo commented on lib/galaxy/tool_util/cwl/util.py:
```diff
@@ -148,12 +155,10 @@ def galactic_job_json(
-        assert "outputs" in upload_response, upload_response
-        assert len(upload_response["outputs"]) > 0, upload_response
-        dataset = upload_response["outputs"][0]
-        datasets.append(dataset)
```
This looks suspicious, without this `datasets` is returned empty but it's used in some places by the callers.

mvdbeek committed:
Implement downloading directory archive

Just get the whole directory archive for now

We could get fancier and get individual files one by one ... but not
sure there's any point in making it so complicated ?

mvdbeek committed:
Quickfix for moving away tool working directory

We should probably just move the contents though ??

mvdbeek committed:
Various fixes for stricter cwltool and cwltest

mvdbeek committed:
Fix up ontology to datatype mapping for __FETCH_DATA__

mvdbeek committed:
Shortcut param_dict building for CWL tools

mvdbeek committed:
WIP: untar directory to extra_files_path

mvdbeek committed:
Add test for workflow default file overrides tool default file

mvdbeek committed:
WIP:CWL default file value_from work

mvdbeek committed:
Into split trans to app

mvdbeek committed:
Separate and fix value_from overriding default

and document hacks. This might not be quite right but all default tests
seem to pass.

mvdbeek committed:
Ensure that expression tool null values are treated as null values when replacing with default

mvdbeek committed:
Replace file location with URL ...

To make that useful we should probably upload deferred datasets and
allow referring to deferred dataset in location scheme.

mvdbeek committed:
Pack workflow

mvdbeek committed:
Update list of new failing 1.2 tests

mvdbeek committed:
Drop now passing red tests

mvdbeek committed:
Exclude red and required 1.0 tests from github matrix

mvdbeek committed:
Fix output addition to history if input name is same as output name

CWL output names are namespaced, so in that case the previous check
doesn't work.

nsoranzo committed:
Avoid unnecessarily recreating ToolProxy for CWL workflow tools

Add ``create_tool_from_proxy()`` method and refactor.

nsoranzo committed:
Avoid unnecessarily recreating ToolProxy when loading CWL tool in toolbox

nsoranzo committed:
Allow dev versions of CWL

Otherwise do_update (which we want to update from older CWL versions)
will update tools to the latest version, which may be a dev one,
causing errors like:

```
schema_salad.exceptions.ValidationException: Version 'v1.3.0-dev1' is a development or deprecated version.
 Update your document to a stable version (v1.0, v1.1, v1.2) or use --enable-dev to enable support for development and deprecated versions.
```

Together with the fixes to not unnecessarily recreate ToolProxy, fix
https://github.com/common-workflow-lab/galaxy/issues/147
